File: java_various_utils/LogsDeleter.java
// NOTE: These helper files are intended to be copied into other projects.
// Change the package `java_various_utils` below to match your project's package
// before integrating.
//
// Recursively scans a project starting from a resolved root directory and deletes
// target folders matching the specified names (e.g. "__pycache__", "logs", "bin",
// "obj", "build", "dist").
//
// HOW IT WORKS
//   1. Directory climbing: starting from a given path, climbs up parent directories
//      until it finds a folder whose name matches ROOT_FOLDER (case-insensitive).
//      This keeps the cleanup confined to the project root and prevents leaking out
//      into outer folders (like C:\ or the user directory).
//   2. Recursive traversal: from that matched root, walks downward and deletes any
//      directory whose name matches one of FOLDERS_TO_REMOVE (children before parents).
//
// No external dependencies (pure JDK).
//
// HOW TO TEST (standalone)
//   javac LogsDeleter.java
//   java LogsDeleter
//   -> climbs up from the working directory to the folder named ROOT_FOLDER and
//      deletes the configured target folders. Edit ROOT_FOLDER before running.

// Change package based on whatever project is implemented
package java_various_utils;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Arrays;
import java.util.Comparator;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public final class LogsDeleter {

    // FORCE USER TO SET THE ROOT FOLDER NAME (case-insensitive),
    // e.g. "drone_battery_performance_prediction"
    public static final String ROOT_FOLDER = "python-various";
    public static final String[] FOLDERS_TO_REMOVE =
            { "__pycache__", "logs", ".pytest_cache", "build", "dist", "bin", "obj" };

    private LogsDeleter() { }

    /**
     * Climbs up directories starting from startPath until it finds a directory
     * whose folder name matches targetName (case-insensitive).
     */
    public static Path findProjectRootByName(String startPath, String targetName) {
        Path current = Paths.get(startPath).toAbsolutePath().normalize();
        if (Files.isRegularFile(current)) {
            current = current.getParent();
        }

        String targetLower = targetName.trim().toLowerCase();
        while (current != null) {
            Path name = current.getFileName();
            if (name != null && name.toString().toLowerCase().equals(targetLower)) {
                return current;
            }
            Path parent = current.getParent();
            if (parent == null || parent.equals(current)) { // reached filesystem root
                break;
            }
            current = parent;
        }

        throw new RuntimeException("Could not find any parent directory named '"
                + targetName + "' starting from '"
                + Paths.get(startPath).toAbsolutePath().normalize() + "'.");
    }

    /**
     * Removes folders matching foldersToRemove from rootDir and all its subdirectories.
     */
    public static void removeFolders(Path rootDir, String[] foldersToRemove) {
        if (!Files.exists(rootDir)) {
            throw new RuntimeException("The specified root directory does not exist: " + rootDir);
        }

        List<String> targets = (foldersToRemove == null || foldersToRemove.length == 0)
                ? Arrays.asList("__pycache__", "logs")
                : Arrays.asList(foldersToRemove);

        List<Path> matches;
        try (Stream<Path> walk = Files.walk(rootDir)) {
            matches = walk
                    .filter(Files::isDirectory)
                    .filter(p -> {
                        Path name = p.getFileName();
                        return name != null && targets.stream()
                                .anyMatch(t -> t.equalsIgnoreCase(name.toString()));
                    })
                    // Delete deepest paths first so nested targets go before their parents
                    .sorted(Comparator.comparingInt(LogsDeleter::depth).reversed())
                    .collect(Collectors.toList());
        } catch (IOException e) {
            System.out.println("Failed to scan " + rootDir + ": " + e.getMessage());
            return;
        }

        for (Path dir : matches) {
            try {
                deleteRecursively(dir);
                System.out.println("Deleted: " + dir);
            } catch (IOException e) {
                System.out.println("Failed to delete " + dir + ": " + e.getMessage());
            }
        }
    }

    private static int depth(Path p) {
        return p.getNameCount();
    }

    private static void deleteRecursively(Path dir) throws IOException {
        try (Stream<Path> walk = Files.walk(dir)) {
            walk.sorted(Comparator.reverseOrder())
                .forEach(p -> {
                    try {
                        Files.delete(p);
                    } catch (IOException e) {
                        throw new RuntimeException(e);
                    }
                });
        } catch (RuntimeException e) {
            if (e.getCause() instanceof IOException) throw (IOException) e.getCause();
            throw e;
        }
    }

    // Main execution entry point.
    public static void main(String[] args) {
        if (ROOT_FOLDER == null || ROOT_FOLDER.isEmpty()) {
            System.out.println("Error: ROOT_FOLDER is not set! Please open this file and specify "
                    + "your project's root folder name in the 'ROOT_FOLDER' constant.");
            return;
        }

        String startPath = System.getProperty("user.dir");
        Path actualRoot = findProjectRootByName(startPath, ROOT_FOLDER);
        System.out.println("Project root found at: " + actualRoot);
        removeFolders(actualRoot, FOLDERS_TO_REMOVE);
    }
}

File: templates/csharp_rest_api_template/src/RestApiTemplate/Utils/LogsDeleter.cs
// NOTE: These helper files are intended to be copied into other projects.
// Change the namespace `cs_various_utils` below to match your project's namespace
// before integrating.
using System;
using System.IO;
using System.Linq;

// Change namespace based on whatever project is implemented
namespace Template.RestApi.Utils
{
    public static class LogsDeleter
    {
        public static string FindProjectRootByName(
            string startPath,
            string? targetName = null
            ){
            var current = Path.GetFullPath(startPath);
            if (File.Exists(current)) current = Path.GetDirectoryName(current);
            var targetLower = targetName?.Trim().ToLower();
            while (!string.IsNullOrEmpty(current))
            {
                if (Path.GetFileName(current).ToLower() == targetLower) return current;
                var parent = Path.GetDirectoryName(current);
                if (string.IsNullOrEmpty(parent) || parent == current) break;
                current = parent;
            }
            throw new Exception($"Could not find a parent directory named '{targetName}' starting from '{startPath}'");
        }

        public static void RemoveLogFolders(
            string rootDir, 
            string[] foldersToRemove
            ){
            var targets = (foldersToRemove == null || foldersToRemove.Length == 0)
                ? new[] { "bin", "obj", "logs" }
                : foldersToRemove;

            if (!Directory.Exists(rootDir)) throw new Exception($"The specified root directory does not exist: {rootDir}");

            // Collect matching directories first, then delete children before parents
            var matches = Directory.EnumerateDirectories(rootDir, "*", SearchOption.AllDirectories)
                .Where(d => targets.Any(t => string.Equals(Path.GetFileName(d), t, StringComparison.OrdinalIgnoreCase)))
                .ToList();

            // Order by depth (longer paths first) so nested targets are deleted before their parents
            var ordered = matches.OrderByDescending(d => d.Count(c => c == Path.DirectorySeparatorChar));

            foreach (var dir in ordered)
            {
                try
                {
                    Directory.Delete(dir, true);
                    Console.WriteLine($"Deleted: {dir}");
                }
                catch (Exception e)
                {
                    Console.WriteLine($"Failed to delete {dir}: {e.Message}");
                }
            }
        }

    }
}
